File: Python/6-LinkedList/Algorithms/1-Floyd’sCycleDetection(Tortoise-and-Hare)/Floyd’sCycleDetection(Tortoise-and-Hare)Pro.py
class ListNode:
    def __init__(self, value=0, next=None):
        """
        Initializes a ListNode object.

        Args:
            value: The value to store in the node (default: 0).
            next:  A reference to the next node in the list (default: None).
                   If None, this is the last node.
        """
        self.value = value  # Store the data
        self.next = next    # Store the reference to the next node

# Floyd’s Cycle Detection Algorithm (Tortoise and Hare)
def detect_cycle_floyd(head):
    """
    Detects if a linked list has a cycle using Floyd's Tortoise and Hare algorithm.

    This algorithm uses two pointers, 'slow' and 'fast'.  The slow pointer moves one
    node at a time, while the fast pointer moves two nodes at a time.  If there's a
    cycle, the two pointers will eventually meet.

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle exists, False otherwise.
    """
    if not head:  # Empty list cannot have a cycle
        return False

    slow, fast = head, head  # Initialize both pointers to the head
    # Iterate until fast reaches the end of the list or a cycle is found
    while fast and fast.next:
        slow = slow.next       # Slow pointer moves one step
        fast = fast.next.next  # Fast pointer moves two steps
        if slow == fast:
            return True  # Cycle detected: slow and fast pointers meet
    return False  # No cycle: fast pointer reached the end of the list

# HashSet Approach
def detect_cycle_hashset(head):
    """
    Detects if a linked list has a cycle using a HashSet (set in Python).

    This approach keeps track of visited nodes in a set. If a node is visited again,
    it means there's a cycle.

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle exists, False otherwise.
    """
    visited = set()  # Create an empty set to store visited nodes
    current = head   # Start at the head of the list
    while current:
        if current in visited:
            return True  # Cycle detected: current node is already in the visited set
        visited.add(current)  # Add the current node to the visited set
        current = current.next  # Move to the next node
    return False  # No cycle: the loop finished without finding a duplicate node

# Modify Node Structure (Flagging Nodes)
def detect_cycle_flagging(head):
    """
    Detects if a linked list has a cycle by modifying the node structure.

    This method changes the 'next' pointer of each visited node to a special 'marker'
    node. If we encounter a node whose 'next' pointer is the marker, we've found a cycle.
    WARNING: This modifies the original linked list.

    Args:
        head: The head node of the linked list.

    Returns:
        True if a cycle exists, False otherwise.
    """
    if not head:
        return False
    marker = ListNode("marker")  # Create a marker node
    current = head             # Start at the head
    while current:
        if current.next == marker:
            return True  # Cycle detected: next pointer points to the marker
        temp = current.next  # Store the next node
        current.next = marker  # Change next pointer to the marker
        current = temp         # Move to the next node (which is now stored in temp)
    return False  # No cycle: the loop finished without finding the marker

# Using Length of Loop (Mathematical Approach)
def detect_cycle_length(head):
    """
    Detects a cycle and returns the length of the cycle if found.

    This approach first uses the slow/fast pointer method to detect a cycle. If a cycle
    is found, it calculates the length of the cycle.

    Args:
        head: The head node of the linked list.

    Returns:
        The length of the cycle if found, 0 otherwise (if no cycle).
    """
    slow, fast = head, head  # Initialize slow and fast pointers
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return cycle_length(slow)  # Call cycle_length to get the length
    return 0  # No cycle

def cycle_length(node):
    """
    Calculates the length of a cycle in a linked list, given a node within the cycle.

    Args:
        node: A node that is part of the cycle.

    Returns:
        The length of the cycle.
    """
    temp = node
    length = 1  # Initialize length to 1 (we've already seen one node)
    while temp.next != node:  # Iterate until we get back to the starting node
        temp = temp.next
        length += 1
    return length

# Using Two Pointers (Variation with Entry Detection)
def detect_cycle_entry(head):
    """
    Detects a cycle and finds the entry point of the cycle.

    This method uses the slow/fast pointer approach.  If a cycle is detected, it finds
    the node where the cycle begins.

    Args:
        head: The head node of the linked list.

    Returns:
        The entry node of the cycle if found, None otherwise.
    """
    if not head:
        return None
    slow, fast = head, head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            break # Exit the loop when slow and fast meet
    else:
        return None  # No cycle: fast reached the end without meeting slow

    slow = head       # Reset slow to the head
    while slow != fast: # Move both pointers one step at a time until they meet
        slow = slow.next
        fast = fast.next
    return slow  # The meeting point is the entry point of the cycle

# Recursive Approach (Less Common)
def detect_cycle_recursive(node, visited=set()):
    """
    Detects a cycle in a linked list using recursion.

    This approach recursively traverses the list, keeping track of visited nodes in a set.
    If a node is visited again, it indicates a cycle.

    Args:
        node:    The current node being visited.
        visited: A set of already visited nodes (default: an empty set).

    Returns:
        True if a cycle exists, False otherwise.
    """
    if node is None:
        return False  # Base case: end of the list, no cycle
    if node in visited:
        return True  # Cycle detected: current node was already visited
    visited.add(node)  # Add the current node to the visited set
    return detect_cycle_recursive(node.next, visited)  # Recursive call for the next node

# Helper function to create a linked list with a cycle for testing
def create_cycle_list(values, cycle_pos):
    """
    Creates a linked list, optionally with a cycle, for testing purposes.

    Args:
        values:     A list of values to create the linked list nodes with.
        cycle_pos:  The position (0-based index) in the list where the cycle should start.
                    If -1, no cycle is created.

    Returns:
        The head node of the created linked list.
    """
    if not values:
        return None  # Empty list
    head = ListNode(values[0])  # Create the head node
    current = head
    nodes = [head]  # Store all created nodes in a list
    for val in values[1:]:
        new_node = ListNode(val)
        nodes.append(new_node)
        current.next = new_node
        current = new_node

    if cycle_pos != -1:
        current.next = nodes[cycle_pos]  # Create the cycle
    return head

# Test Cases
values = [1, 2, 3, 4, 5]
cycle_head = create_cycle_list(values, 2)  # Create a list with a cycle starting at node with value 3
non_cycle_head = create_cycle_list(values, -1) # Create a list without a cycle

print("Floyd’s Cycle Detection:", detect_cycle_floyd(cycle_head))
print("HashSet Approach:", detect_cycle_hashset(cycle_head))
print("Flagging Approach:", detect_cycle_flagging(cycle_head))
print("Cycle Length:", detect_cycle_length(cycle_head))
entry_node = detect_cycle_entry(cycle_head)
print("Cycle Entry Node:", entry_node.value if entry_node else None)
print("Recursive Detection:", detect_cycle_recursive(cycle_head))

File: Java/6-LinkedList/Algorithms/1-Floyd’sCycleDetection(Tortoise-and-Hare)/Floyd’sCycleDetection(Tortoise-and-Hare)Pro.java
import java.util.HashSet;
import java.util.Set;
import java.util.HashMap;
import java.util.Map;

public class LinkedListCycleDetection {

    // Definition of a singly linked list node.
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }

    public static void main(String[] args) {
        // Create a sample linked list with a cycle for testing.
        ListNode head = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);

        head.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node2; // Creating a cycle: node5 points back to node2.

        // Test the different cycle detection methods.
        System.out.println("Floyd's Cycle Detection (Tortoise and Hare): " + hasCycleFloyd(head)); // Expected: true
        System.out.println("Using HashSet: " + hasCycleUsingHashSet(head)); // Expected: true
        System.out.println("Using HashMap: " + hasCycleUsingHashMap(head)); // Expected: true
        System.out.println("Modifying List Nodes: " + hasCycleModifyingList(head)); // Expected: true, but modifies the list!
        //Recreate the list
        head = new ListNode(1);
        node2 = new ListNode(2);
        node3 = new ListNode(3);
        node4 = new ListNode(4);
        node5 = new ListNode(5);

        head.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node2;
        System.out.println("Using Length Comparison: " + hasCycleLengthComparison(head));
    }

    // 1. Floyd's Cycle Detection Algorithm (Tortoise and Hare)
    //   - Uses two pointers, 'slow' and 'fast'.
    //   - 'slow' moves one step at a time, 'fast' moves two steps at a time.
    //   - If there is a cycle, the two pointers will eventually meet.
    //   - This is the most efficient and commonly used algorithm.
    //
    //   - Real-world application: Detecting loops in program execution paths,
    //     finding circular dependencies in data structures, and optimizing
    //     memory management by identifying unreachable objects.
    public static boolean hasCycleFloyd(ListNode head) {
        if (head == null || head.next == null) {
            return false; // No cycle if the list is empty or has only one node.
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;      // Move slow pointer one step.
            fast = fast.next.next; // Move fast pointer two steps.

            if (slow == fast) {
                return true; // Cycle detected: slow and fast pointers meet.
            }
        }
        return false; // No cycle: fast pointer reached the end of the list.
    }

    // 2. Using a HashSet
    //   - Stores each visited node in a HashSet.
    //   - If a node is encountered again, it means there is a cycle.
    //   - Uses extra space for the HashSet.
    //
    //   - Real-world application: Detecting duplicate entries in data streams,
    //     identifying previously visited states in state machines, and
    //     preventing infinite loops in web crawling algorithms.
    public static boolean hasCycleUsingHashSet(ListNode head) {
        if (head == null) {
            return false;
        }

        Set<ListNode> visitedNodes = new HashSet<>();
        ListNode current = head;

        while (current != null) {
            if (visitedNodes.contains(current)) {
                return true; // Cycle detected: current node is already in the set.
            }
            visitedNodes.add(current); // Add the current node to the set.
            current = current.next;     // Move to the next node.
        }
        return false; // No cycle: reached the end of the list.
    }

    // 3. Using a HashMap
    //   - Stores each visited node and its count in a HashMap.
    //   - If a node is encountered again, the count will be greater than 1,
    //     indicating a cycle.
    //   - Similar to HashSet but stores counts, which can be useful in other scenarios.
    //
    //  - Real-world application: Counting the frequency of visited URLs in
    //    web crawling, tracking the number of times a specific process
    //    is executed, and analyzing network traffic patterns to detect
    //    repeatedly visited nodes.
    public static boolean hasCycleUsingHashMap(ListNode head) {
        if (head == null) {
            return false;
        }

        Map<ListNode, Integer> nodeCounts = new HashMap<>();
        ListNode current = head;

        while (current != null) {
            if (nodeCounts.containsKey(current)) {
                return true; // Cycle detected: node count is greater than 1.
            }
            nodeCounts.put(current, 1); // Add the node to the map or increment its count.
            current = current.next;       // Move to the next node.
        }
        return false; // No cycle: reached the end of the list.
    }

    // 4. Modifying List Node Values (Not Recommended for Production)
    //   - Changes the 'val' field of each visited node to a specific value.
    //   - If a node with the specific value is encountered again, it indicates a cycle.
    //   - **Caution:** This method modifies the original linked list, which is
    //     generally not recommended.  It's destructive and can lead to problems
    //     if the list is used elsewhere.
    //   - It also only works for specific data types and value ranges.
    //
    //   - Real-world application: This approach is generally *not* recommended
    //     for real-world applications due to its destructive nature.  It might
    //     be used in very specific, isolated scenarios where modifying the
    //     list is acceptable and memory efficiency is extremely critical, but
    //     alternatives are almost always preferred.  It's included here for
    //     completeness and to illustrate a less desirable approach.
    public static boolean hasCycleModifyingList(ListNode head) {
        if (head == null) {
            return false;
        }

        ListNode current = head;
        int markerValue = Integer.MIN_VALUE; // Use a value that's unlikely to appear in the list.

        while (current != null) {
            if (current.val == markerValue) {
                return true; // Cycle detected: node value has been seen before.
            }
            current.val = markerValue; // Mark the node as visited.
            current = current.next;     // Move to the next node.
        }
        return false; // No cycle: reached the end of the list.
    }

    // 5. Length Comparison (Only Works for Specific Cycle Types)
    //   - Calculate the length of the list by traversing it.
    //   - If a cycle exists, the traversal might not terminate, or a very long traversal.
    //   - This method is unreliable for general cycle detection, especially for complex cycles.
    //   - It's more of a conceptual approach and not robust.
    //   - In practical scenarios, you might set a maximum allowed length to prevent infinite loops.
    //
    //   - Real-world application: This method is generally *not* suitable for
    //     real-world applications due to its limitations and potential for
    //     infinite loops.  It might be used in very specific scenarios where
    //     the cycle is known to be a simple loop at the end of the list, and
    //     a rough estimate of length is needed.  However, Floyd's algorithm
    //     or HashSet are far more reliable and efficient.
    public static boolean hasCycleLengthComparison(ListNode head) {
        if (head == null) {
            return false;
        }

        int originalLength = 0;
        ListNode current = head;
        while (current != null && originalLength < 200) { //Added a max length check
            originalLength++;
            current = current.next;
        }
        if(current == null){
            return false;
        }

        int traversedLength = 0;
        current = head;
        while (current != null && traversedLength <= originalLength) {
            traversedLength++;
            current = current.next;
        }

        return traversedLength > originalLength;
    }
}

